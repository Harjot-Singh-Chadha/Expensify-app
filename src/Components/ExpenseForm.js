import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

const now = moment();
console.log(now.format("MMM Do YYYY h:mm:ss A"));

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      notes: props.expense ? props.expense.notes : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calenderFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    this.setState({
      description: e.target.value
    });
  };

  onNotesChange = e => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  };

  onAmountChange = e => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: "Please enter description and amount" }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        notes: this.state.notes
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          type="text"
          placeholder="Description"
          className="text-input"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          className="text-input"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />

        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={this.onDateChange} // PropTypes.func.isRequired
          focused={this.state.calenderFocused} // PropTypes.bool
          onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
          numberOfMonths={1}
          isOutsideRange={() => {
            false;
          }}
        />

        <textarea
          className="textarea"
          placeholder="Add a note for your expense {Optional}"
          value={this.state.notes}
          onChange={this.onNotesChange}
        ></textarea>
        <div>
          <button className="button" onClick={this.onSubmit}>
            Save Expense
          </button>
        </div>
      </form>
    );
  }
}
