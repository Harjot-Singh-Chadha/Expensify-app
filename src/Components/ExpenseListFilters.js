import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      calendarFocused: null,
      startDate: null,
      endDate: null
    };
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.setState({
      startDate,
      endDate
    });
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calendarFocused => {
    this.setState({ calendarFocused });
  };

  onTextChange = e => {
    this.props.setTextFilter({ text: e.target.value });
  };

  onSortChange = e => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container ">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search Expenses"
              type="text"
              value={this.props.filters.text}
              onChange={this.onTextChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            {this.props.filters.sortBy === "date" && (
              <DateRangePicker
                startDate={
                  this.state.startDate === null
                    ? this.state.startDate
                    : this.props.filters.startDate
                } // momentPropTypes.momentObj or null,
                startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                endDate={
                  this.state.endDate === null
                    ? this.state.endDate
                    : this.props.filters.endDate
                } // momentPropTypes.momentObj or null,
                endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                focusedInput={this.state.calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => {
                  false;
                }}
              />
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
