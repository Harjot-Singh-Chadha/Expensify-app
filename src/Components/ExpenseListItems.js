import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
import numeralCurrencyFormat from "../locales/fr";

const ExpenseListItems = ({ description, amount, createdAt, id }) => (
  <Link to={`./edit/${id}`} className="list-item">
    <div className="list-item__wrapper">
      <h3 className="list-item__title">{description}</h3>
      <br />
      <span className="list-item__sub-title">
        {moment(createdAt).format("MMMM Do YYYY")}
      </span>
    </div>
    <h3 className="list-item__data">
      {numeral(amount / 100).format("$0,0.00")}
    </h3>
  </Link>
);

export default ExpenseListItems;
