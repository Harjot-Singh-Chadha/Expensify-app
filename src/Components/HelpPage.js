import React from 'react';

const HelpPage = () => (

    <div>This is a HelpPage</div>
);

export default HelpPage