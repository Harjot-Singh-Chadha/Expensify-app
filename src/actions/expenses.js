import database from '../firebase/firebase';

// ADD_EXPENSE

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
     return (dispatch) => {
        const {
            description = '',
            notes = '',
            amount = 0,
            createdAt = 0 
        } = expenseData;
        const expense = { description, notes, amount, createdAt };
       
      return database.ref('expenses')
        .push(expense)
        .then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        .catch((err)=>{
            console.log(err);
        })
        });
     };

};

// REMOVE_EXPENSE

export const removeExpense = ({id} = {}
) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id } = {}) =>{
  
    return (dispatch) => {
      
   return database.ref(`expenses/${id}`).remove().then(()=>{
          dispatch(removeExpense({id}))
     });

    }

};

// EDIT_EXPENSE

export const editExpense = (id, updates) => ({
        type: 'EDIT_EXPENSE',
        id,
        updates
    });

 export const startEditExpense = (id, updates) =>{
  
        return (dispatch) => {
        
            return database.ref(`expenses/${id}`).update(updates).then(()=>{
              
              dispatch(editExpense(id, updates))
         });
    
        }
    
    };

//Set_Expenses

export const setExpenses = (expenses) => ({
    type: 'Set_Expenses',
    expenses
});

export const startSetExpense = () => {
     
    return (dispatch) => {
      
     return database.ref('expenses')
    .once('value')
    .then((DataSnapshot)=>{

      const expenses = [] ;
       DataSnapshot.forEach((childSnapshot)=>{
          expenses.push({
            id:childSnapshot.key,
            ...childSnapshot.val()
          })
       })
      dispatch(setExpenses(expenses));
    })
    .catch((err) =>{
         console.log(`Error getting data from the DB`,err);
    })
    
};

};

    