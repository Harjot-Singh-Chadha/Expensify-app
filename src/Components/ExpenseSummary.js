import React from "react";
import Modal from "react-modal";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import SelectExpensesTotal from "../selectors/expenses-total";
import { startRemoveAllExpense } from "../actions/expenses";
import SelectExpenses from "../selectors/expenses";
import numeral from "numeral";

export class ExpenseSummary extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      modalIsOpen: false
    };
  }
  openModal = () => {
    if (this.props.expensesCount !== 0) {
      this.setState({ modalIsOpen: true });
    }
  };

  closeModal = () => {
    this.setState({ modalIsOpen: false });
  };

  onRemoveAll = () => {
    this.props.startRemoveAllExpense();
    this.closeModal();
  };

  render() {
    return (
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">
            Viewing <span> {this.props.expensesCount} </span>
            {this.props.expensesCount === 1 ? "expense" : "expenses"} totalling
            <span>
              {" "}
              {numeral(this.props.expensesTotal / 100).format("$0,0.00")}{" "}
            </span>
          </h1>
          <div className="page-header__actions">
            <Link className="button" to="/create">
              Add Expense
            </Link>
            <button className="button-removeAll" onClick={this.openModal}>
              Remove All
            </button>
            <Modal
              className="modal"
              isOpen={this.state.modalIsOpen}
              onRequestClose={this.closeModal}
              ariaHideApp={false}
              contentLabel="Example Modal"
              closeTimeoutMS={200}
            >
              <div className="modal-content__wrapper">
                <h1 className="modal-content__title">Are you sure</h1>
                <div className="modal-content-button__wrapper">
                  <button className="button" onClick={this.onRemoveAll}>
                    Yes
                  </button>
                  <button className="button" onClick={this.closeModal}>
                    No
                  </button>
                </div>
              </div>
            </Modal>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const visibleExpenses = SelectExpenses(state.expenses, state.filters);
  return {
    expensesTotal: SelectExpensesTotal(visibleExpenses),
    expensesCount: visibleExpenses.length
  };
};

const mapDispatchToProps = dispatch => ({
  startRemoveAllExpense: () => dispatch(startRemoveAllExpense())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseSummary);
