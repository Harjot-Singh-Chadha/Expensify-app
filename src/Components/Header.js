import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (
  <header className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Expensify</h1>
        </Link>
        &nbsp;
        {/* <NavLink to='/create' activeClassName='is-active'>Add Expense</NavLink>&nbsp; */}
        <div className="">
          <button className=" button-header__logout" onClick={startLogout}>
            Logout
          </button>
        </div>
      </div>
    </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
