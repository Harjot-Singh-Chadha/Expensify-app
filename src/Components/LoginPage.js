import React from 'react';
import { connect } from 'react-redux';
import { startGoogleLogin, startFaceBookLogin } from  '../actions/auth';

export const LoginPage = ({startGoogleLogin, startFaceBookLogin}) => (

  <div>
      <h1>Login here</h1>
      <button onClick={startGoogleLogin} >Google</button>
      <button onClick={startFaceBookLogin} >FaceBook</button>
  </div>
    

);

const mapDispatchToProps = (dispatch) => ({

  startGoogleLogin: () => dispatch(startGoogleLogin()),
  startFaceBookLogin: () => dispatch(startFaceBookLogin())

});

export default connect(undefined, mapDispatchToProps)(LoginPage);