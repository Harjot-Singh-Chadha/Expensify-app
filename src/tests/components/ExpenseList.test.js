import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../Components/ExpenseList';
import expenses from '../fixtures/expenses';

test("Should render expense list with expense items", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
 
});

test("Should render expense list with empty message", () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);
    expect(wrapper).toMatchSnapshot();
   
  });