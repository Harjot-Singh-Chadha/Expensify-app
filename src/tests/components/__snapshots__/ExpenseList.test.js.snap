// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className=" list-item list-item__message"
    >
      <span>
        No Expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`Should render expense list with expense items 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItems
      amount={195}
      createdAt={"1970-01-01T00:00:00.000Z"}
      description="Gum"
      id="1"
      key="1"
      notes=""
    />
    <ExpenseListItems
      amount={19500}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      notes=""
    />
    <ExpenseListItems
      amount={4500}
      createdAt={345600000}
      description="Credit card"
      id="3"
      key="3"
      notes=""
    />
  </div>
</div>
`;
